### ISSUE TITLE
ID extractor does not seems to be working

### ISSUE THREAD

**Issue by buger** [TykTechnologies/tyk#1380]
**Do you want to request a *feature* or report a *bug*?**
bug
**What is the current behavior?**
While it generates the token representation, and store it in the cache, I was never able to actually get it return cached value, in other words, requests always get passed to the auth middleware.
**What is the expected behavior?**
ID extractor is intended to cache auth middleware response and after successful auth, authenticate the user from the cache, bypassing auth middleware.
**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**
It is quite easy to replicate, personally, I used the following middleware:
```python
from tyk.decorators import *
from gateway import TykGateway as tyk

counter = 0

@Hook
def MyAuthHook(request, session, metadata, spec):
    print("MyAuthHook is called")
    global counter
    counter = counter + 1
    auth_header = request.get_header('Authorization')
    if auth_header == 'valid_token' and counter < 2:
        session.rate = 1000.0
        session.per = 1.0
        metadata["token"] = "valid_token"
    return request, session, metadata
```

and manifest

```json
		{
		    "file_list": [
		        "middleware.py"
		    ],
		    "custom_middleware": {
		        "driver": "python",
		        "auth_check": {
		            "name": "MyAuthHook"
		        },
		        "id_extractor": {
		        	"extract_from": "header",
		        	"extract_with": "value",
		        	"extractor_config": {
		        		"header_name": "Authorization"
		        	}
		        }
		    }
		}
```

The basic idea that it uses global counter and restrict running it more the 1 time. When ID extractor turned on, if auth is successful, you should be able to make multiple requests to protected API. If ID extractor cache not working, it should work only on the first call and give you auth error on the following requests.

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**
master (maybe older versions too)

**Comment by buger** [comment]
Looking at the code I can see 2 issue, and not quite understand how it worked before. 

Extractor implementations have `ExtractAndCheck` method, which should return `session_id` and `returnOverrides` object. `returnOverrides.ResponseCode` set to 200 it indicates that cached value is found, and we can bypass auth plugin. 

All the `ExtractAndCheck` functions have the same check:
```go
if keyExists {
		lastUpdated, _ := strconv.Atoi(previousSession.LastUpdated)
		deadlineTs := int64(lastUpdated) + previousSession.IdExtractorDeadline

		log.Error("Previous session:", previousSession.LastUpdated)

		if deadlineTs > time.Now().Unix() {
			e.PostProcess(r, &previousSession, SessionID)
			returnOverrides = ReturnOverrides{
				ResponseCode: 200,
			}
		}
	}
```

To make it work `deadlineTs` should be non zero and bigger than the current time. In my tests, it is always zero, which means both `previousSession.LastUpdated` and `previousSession.IdExtractorDeadline` are zero.

Looking at the code `LastUpdated` field is set in 2 general cases:
1. When the key is created via API or inherited from Policy.LastUpdated field. In case of ID extractor,
2. When we call `CheckSessionAndIdentityForValidKey`, key not found in `keystore` but found in `authstore`. In on-premise and community versions `keystore` and `authstore` are equal (redis), so it never actually use `authstore` clause.  For the Hybrid installation, `authstore` is separate storage (calling RPC). 

Both of the cases does not apply to id extractor because Session object created dynamically based on auth middleware response and I can't find a place which sets `LastUpdated` field anywhere. In addition `CheckSessionAndIdentityForValidKey` does not get called for dynamic sessions too. 

Regarding `IdExtractorDeadline`, which is required to make id extractor work, there is no placed in the code which set this value anywhere. So I assume this value should come from the plugin, but it does not mention anywhere. In fact, this value is just an ID extractor session timeout. 

Considering all this, since session object can be modified inside plugin, I modified my plugin to make it work by adding following lines:
```
        session.last_updated = str(int(time.time()))
        session.id_extractor_deadline = 60
```
Basically manually setting fields mentioned above.

So, my proposal is 
1. Document that setting `id_extractor_deadline` is required, should specify ID extractor cache TTL, and should be set inside Auth plugin
2. Set session.LastUpdated field when creating an id extractor cached session.

@matiasinsaurralde does it looks sane?

**Comment by lonelycode** [comment]
> Document that setting id_extractor_deadline is required, should specify ID extractor cache TTL, and should be set inside Auth plugin

Yes this is a good move

> Set session.LastUpdated field when creating an id extractor cached session

Or document that the auth MW should do this when it passes the session object back.

**Comment by matiasinsaurralde** [comment]
> Regarding IdExtractorDeadline, which is required to make id extractor work, there is no placed in the code which set this value anywhere. So I assume this value should come from the plugin, but it does not mention anywhere. In fact, this value is just an ID extractor session timeout.

Do you think we should introduce a default value here (could be part of a different ticket)?

**Comment by buger** [comment]
Ok, I found the issue, everything works as expected, and our docs mentioned it the right way: `session.id_extractor_deadline = timestamp + 60`. 

I was confused for multiple reasons, first of all `LastUpdated` field is used inside id_extractor and it always zero, and by looking at the code, it feels like deadline should be a relative value, but in fact it is absolute. 

I fixed the tests to follow docs https://github.com/TykTechnologies/tyk/commit/cb31ed781b26d6001b225635be523b703ba3d6f5

And removed usage of LastUpdated field, as it never gets set https://github.com/TykTechnologies/tyk/pull/1400

**Comment by lonelycode** [comment]
Can the value still be set in the MW though? The field is to prevent unnecessary writes to redis for quota updates iirc?

**Comment by buger** [comment]
It can be set via middleware, but it no longer used in ID extractor logic. 

Moreover, I think there was a bug there since the on-premise version was never setting last-updated field, and hybrid version was setting it on each key retrieval from RPC layer. So in some cases, it was used in id extractor ttl calculations, in some not.  So I think it is true to say that id_extractor was broken for hybrid/mdcb installations.
